feat[CI]: Add golangci-lint

iamrajiv opened the pull request:
Fixes: #13

mgfeller commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,27 @@
+name: MeshKit
+on:
+  push:
+    branches: [ master ]
```
In order to detect problems and fail early this should also run on pushes to all branches, and I think also tags:

```on:
  push:
    branches:
      - '**'
    tags:
      - 'v*'
  pull_request:
    branches:
      - master
```

mgfeller reviewed (changes requested):
Excellent @iamrajiv thank you!

Please have a look at the inlined comment.

mgfeller commented:
@iamrajiv isn't `.golangci.yml` required in the root directory (as here https://github.com/meshery/meshery-consul/blob/master/.golangci.yml)?  I'd expect running `make check` should catch the e.g. cases where file were not `goimports`-ed and `go fmt`'ed, as now in errorutil code. I always run `make check` and trust it. It used to create a lot of errors here in `meshkit` (hence the issue I created) and suddenly those were not reported anymore. Didn't have time to look into that then, unfortunately....

mgfeller commented:
If I add the aforementioned `.golangci.yml` I get

```
❯ make check
golangci-lint run
utils/kubernetes/service_test.go:32: File is not `gofmt`-ed with `-s` (gofmt)
                                corev1.ServicePort{
cmd/errorutil/internal/error/summary.go:5: File is not `goimports`-ed with -local github.com/golangci/golangci-lint (goimports)
        "github.com/layer5io/meshkit/cmd/errorutil/internal/component"
        log "github.com/sirupsen/logrus"
cmd/errorutil/internal/coder/file.go:5: File is not `goimports`-ed with -local github.com/golangci/golangci-lint (goimports)
        "github.com/layer5io/meshkit/cmd/errorutil/internal/component"
        errutilerr "github.com/layer5io/meshkit/cmd/errorutil/internal/error"
        "github.com/sirupsen/logrus"
cmd/errorutil/internal/coder/node.go:5: File is not `goimports`-ed with -local github.com/golangci/golangci-lint (goimports)
        "github.com/layer5io/meshkit/cmd/errorutil/internal/component"
        "github.com/sirupsen/logrus"
logger/types.go:4:2: const `JsonLogFormat` should be `JSONLogFormat` (golint)
        JsonLogFormat = iota
        ^
utils/network.go:25:6: func TcpCheck should be TCPCheck (golint)
func TcpCheck(hp *HostPort, mock *MockOptions) bool {
     ^
broker/nats/nats.go:118:1: receiver name in should be consistent with previous receiver name n for Nats (golint)
func (in *Nats) DeepCopyInto(out broker.Handler) {
^
broker/nats/nats.go:123:1: receiver name in should be consistent with previous receiver name n for Nats (golint)
func (in *Nats) DeepCopy() *Nats {
^
broker/nats/nats.go:133:1: receiver name in should be consistent with previous receiver name n for Nats (golint)
func (in *Nats) DeepCopyObject() broker.Handler {
^
utils/utils.go:81:15: G107: Potential HTTP request made with variable url (gosec)
        resp, err := http.Get(url)
                     ^
utils/utils.go:145:19: G107: Potential HTTP request made with variable url (gosec)
        response, err := http.Get(url)
                         ^
cmd/errorutil/internal/coder/node.go:118:58: `analyses` is a misspelling of `analyzes` (misspell)
// handleValueSpec inspects node n if it is a ValueSpec, analyses and updates it (depending on update and updateAll). Returns true if any value was changed.
                                                         ^
models/oam/core/v1alpha1/doc.go:9:13: `Definiton` is a misspelling of `Definition` (misspell)
// 5. Scope Definiton
            ^
utils/kubernetes/expose/traverser.go:45:47: `occured` is a misspelling of `occurred` (misspell)
// while the other is the error that may have occured in the traverser
                                              ^
errors/errors.go:6: unnecessary leading newline (whitespace)
func NewDefault(code string, ldescription ...string) *Error {

errors/errors.go:18: unnecessary leading newline (whitespace)
func New(code string, severity Severity, sdescription []string, ldescription []string, probablecause []string, remedy []string) *Error {

errors/errors.go:32: unnecessary leading newline (whitespace)
func GetCode(err error) string {

utils/kubernetes/expose/expose.go:10:2: ST1019: package "k8s.io/api/core/v1" is being imported more than once (stylecheck)
        corev1 "k8s.io/api/core/v1"
        ^
utils/kubernetes/apply-manifest.go:8:2: ST1019: package "k8s.io/apimachinery/pkg/api/errors" is being imported more than once (stylecheck)
        "k8s.io/apimachinery/pkg/api/errors"
        ^
utils/kubernetes/service_test.go:9:2: ST1019: package "k8s.io/api/core/v1" is being imported more than once (stylecheck)
        corev1 "k8s.io/api/core/v1"
        ^
```

That's more like it :+1:

mgfeller reviewed (commented):
@iamrajiv would you mind adding `.golangci.yml` (I only did it locally to test) or how would you suggest going forward?

mgfeller reviewed (changes requested):
@iamrajiv there seems to have been a misunderstanding, sorry for not being clear enough. With `.golangci.yml` I was referring to the config file for `golangci-lint`, the workflow itself should remain in the `.github/workflows` directory.
Adding the config file - and handling the reported errors - could be addressed in a separate PR... what do you think?

iamrajiv commented:
@mgfeller I have reverted the commit. Please check once.

mgfeller reviewed (changes requested):
Thanks! Not quite sure what it was reverted to, but I think it should have been

```
on:
  push:
    branches:
      - '**'
    tags:
      - 'v*'
  pull_request:
    branches:
      - master
```
shouldn't it?

iamrajiv commented:
@mgfeller Yeah I have updated the ci can u check once

mgfeller commented:
@iamrajiv cool it works :smile: there are some warnings even... do you fix them? I can fix them if you like in this PR, they are from some of "my" errorutil code anyway.

iamrajiv commented:
@mgfeller there was an issue with a timeout so I added a configuration file fo `golangci-lint` where I added `timeout: 5m`. I will edit this configuration in separate PR where I will add other settings.
